File: examples/simple_controller.py
#!/usr/bin/env python3
"""
Minimal example: uses an ONNX policy to control Lite3 via the highlevel SDK.

Usage:
    python simple_controller.py [robot_ip] [model_path]

Defaults: robot_ip=192.168.1.2, model_path=policy.onnx
"""

import sys
import os
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "python"))

from lite3_highlevel import Bridge, RobotState, JointCommand
from lite3_highlevel import OnnxRunner, ObsBuilder, CmdProcessor


def main():
    robot_ip   = sys.argv[1] if len(sys.argv) > 1 else "192.168.1.2"
    model_path = sys.argv[2] if len(sys.argv) > 2 else "policy.onnx"

    print(f"=== Lite3 Highlevel Controller ===")
    print(f"Robot IP: {robot_ip}")
    print(f"Model:    {model_path}")

    # Load ONNX policy
    runner = OnnxRunner(model_path)
    builder = ObsBuilder()
    processor = CmdProcessor()

    last_action = [0.0] * 12
    tick = 0

    def policy(state: RobotState) -> JointCommand:
        nonlocal last_action, tick
        tick += 1

        # Build observation from sensor data
        obs = builder.build(state, last_action=last_action)

        # Run ONNX inference
        raw_action = runner.infer(obs)
        last_action = raw_action

        # Convert to joint command
        cmd = processor.process(raw_action)

        if tick % 50 == 0:
            rpy_deg = tuple(v * 57.3 for v in state.rpy)
            print(f"[{tick:5d}] rpy=({rpy_deg[0]:5.1f},{rpy_deg[1]:5.1f},{rpy_deg[2]:5.1f}) deg"
                  f"  cmd_vel=({state.cmd_vel[0]:.2f},{state.cmd_vel[1]:.2f},{state.cmd_vel[2]:.2f})")

        return cmd

    # Run control loop
    bridge = Bridge(robot_ip=robot_ip)
    try:
        bridge.run(policy, frequency_hz=50)
    except KeyboardInterrupt:
        print("\nShutting down...")
        bridge.shutdown()


if __name__ == "__main__":
    main()

File: examples/simple_controller.cpp
/**
 * @file simple_controller.cpp
 * @brief Minimal example: ONNX policy → Lite3 control via highlevel SDK (C++)
 *
 * Demonstrates the handshake protocol:
 *   1. Bridge waits for robot to enter RLHandshakeMode
 *   2. Sends HEARTBEAT_READY until robot enters RLControlMode
 *   3. Runs the policy control loop at 50 Hz
 *
 * Build:
 *   g++ -std=c++17 -I../cpp/include -I<onnxruntime>/include
 *       simple_controller.cpp ../cpp/src/onnx_runner.cpp
 *       -L<onnxruntime>/lib -lonnxruntime -lpthread
 */

#include "lite3_highlevel/lite3_highlevel.hpp"

#include <iostream>
#include <iomanip>
#include <cmath>
#include <csignal>
#include <chrono>
#include <thread>

static volatile bool g_running = true;

void signal_handler(int) { g_running = false; }

int main(int argc, char* argv[]) {
    std::string robot_ip   = (argc > 1) ? argv[1] : "192.168.1.2";
    std::string model_path = (argc > 2) ? argv[2] : "policy.onnx";

    std::signal(SIGINT, signal_handler);

    std::cout << "=== Lite3 Highlevel Controller (C++) ===" << std::endl;
    std::cout << "Robot IP: " << robot_ip << std::endl;
    std::cout << "Model:    " << model_path << std::endl;

    // --- Setup ---
    lite3::OnnxRunner runner(model_path);
    lite3::ObsBuilder builder;
    lite3::CmdProcessor processor;
    lite3::Bridge bridge(robot_ip);

    if (!bridge.open()) {
        std::cerr << "Failed to open bridge" << std::endl;
        return 1;
    }

    float obs[45];
    float raw_action[12];
    float last_action[12] = {0};
    int tick = 0;

    bool policy_initialised = false;  // set up once when handshake completes
    const double period_s = 1.0 / 50.0;  // 50 Hz

    std::cout << "Running at 50 Hz... Press Ctrl+C to stop." << std::endl;

    // --- Control loop with handshake ---
    while (g_running) {
        auto tick_start = std::chrono::steady_clock::now();

        // 1. Receive latest sensor data
        struct SensorPacket pkt;
        bool got = bridge.recv(pkt);

        lite3::RobotMotionState robot_state = lite3::RobotMotionState::WaitingForStand;
        if (got) {
            robot_state = static_cast<lite3::RobotMotionState>(pkt.current_state);
        }

        // 2. Update handshake state machine
        bridge.updateHandshake(robot_state);

        // 3. Act based on handshake state
        auto hs = bridge.getHandshakeState();

        if (hs == lite3::HandshakeState::ESTABLISHED) {
            // --- Normal policy loop ---
            if (!policy_initialised) {
                // Reset observation state on fresh handshake
                std::memset(last_action, 0, sizeof(last_action));
                policy_initialised = true;
            }

            if (got) {
                // Convert to RobotState
                lite3::RobotState state;
                state.seq = pkt.seq;
                state.timestamp_ms = pkt.timestamp_ms;
                state.current_state = pkt.current_state;
                for (int i = 0; i < 3; ++i) {
                    state.rpy[i]   = pkt.rpy[i];
                    state.acc[i]   = pkt.acc[i];
                    state.omega[i] = pkt.omega[i];
                    state.cmd_vel[i] = pkt.cmd_vel[i];
                }
                std::memcpy(state.joint_pos, pkt.joint_pos, sizeof(state.joint_pos));
                std::memcpy(state.joint_vel, pkt.joint_vel, sizeof(state.joint_vel));
                std::memcpy(state.joint_tau, pkt.joint_tau, sizeof(state.joint_tau));

                // Build observation
                builder.setLastAction(last_action);
                builder.build(state, obs);

                // Run ONNX inference
                runner.infer(obs, raw_action);
                std::memcpy(last_action, raw_action, sizeof(last_action));

                // Convert to joint command and send
                auto cmd = processor.process(raw_action);
                bridge.sendCommand(cmd);
            } else {
                // No sensor data — send zero hold (safe)
                bridge.sendZeroHold();
            }

            // Periodic status
            if (got && ++tick % 50 == 0) {
                std::cout << "[" << tick << "]"
                          << " rpy=("
                          << std::fixed << std::setprecision(1)
                          << pkt.rpy[0]*57.3 << ","
                          << pkt.rpy[1]*57.3 << ","
                          << pkt.rpy[2]*57.3 << ") deg"
                          << std::endl;
            }

        } else if (hs == lite3::HandshakeState::HANDSHAKING) {
            // Send HEARTBEAT_READY and wait for RLControlMode
            bridge.sendReady();

        } else {
            // DISCONNECTED — send zero hold (safe while robot is idle/standing up)
            if (got) {
                bridge.sendZeroHold();
            }
            policy_initialised = false;
        }

        // Maintain frequency
        auto elapsed = std::chrono::steady_clock::now() - tick_start;
        auto remaining = std::chrono::duration<double>(period_s) - elapsed;
        if (remaining.count() > 0)
            std::this_thread::sleep_for(remaining);
    }

    std::cout << "\nShutting down..." << std::endl;
    bridge.sendShutdown();
    bridge.close();
    return 0;
}
